### PR TITLE
Python3.10

### PR DESCRIPTION
**Pull request by b-butler**
## Description

Add Python 3.10 support.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [x] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/master/changelog.txt) and added all related issue and pull request numbers for future reference (if applicable). See example below.

### CODE REVIEW

**Comment by csadorf** [comment]
@b-butler Looks like https://github.com/h5py/h5py/pull/1973 that was blocking this PR has been merged.

**Comment by bdice** [comment]
I think this could be blocked for a while longer, until h5py has wheels available for Python 3.10. That will likely involve solving this [separate issue for Mac + Python 3.10 support](https://github.com/h5py/h5py/pull/1987) and h5py publishing those new wheels to PyPI. For now, I think we could adjust the signac tests on 3.10 to run with “minimal dependencies” and merge this PR. We do support Python 3.10, though we have optional dependencies that may take a bit of time to catch up.

**Comment by csadorf** [comment]
> I think this could be blocked for a while longer, until h5py has wheels available for Python 3.10, at least on Linux. Those wheels may not be published until this [separate issue for Mac + Python 3.10 support](https://github.com/h5py/h5py/pull/1987) is solved. For now, I think we could adjust the signac tests on 3.10 to run with “minimal dependencies” (which takes out the dependencies on h5py) and merge this PR. We do support Python 3.10, though we have optional dependencies that may take a bit of time to catch up.

I think it's a bit of a matter interpretation of what "support" means. I would generally think that the full feature set should be supported, but given that it might take a very long time, agree that merging now and then tracking the full support in a different issue and PR is acceptable.

However, especially on conda this could be problematic, because there we typically install with all optional dependencies. I think it would be surprising to the user if the software came with a differing feature set as a function of the selected Python version.

**Comment by bdice** [comment]
> I think it would be surprising to the user if the software came with a differing feature set as a function of the selected Python version.

This might not be an issue. Users already have to install h5py manually if installing from PyPI/source. On conda-forge, the package would fail to install if the solver cannot meet the required h5py dependency for a given version of Python. The signac package is a universal wheel on PyPI and a `noarch` package on conda-forge so users can _already_ install signac if they have Python 3.10 installed. This PR just affects our ability to do CI and "quality assurance," so to speak.

**Comment by csadorf** [comment]
Ok, then I would suggest that we add a note in the changelog entry and the README about the limitation, but otherwise go ahead and merge this.

**Comment by bdice** [comment]
@b-butler It looks like the other Linux tests in CircleCI might inherit the “minimal” dependencies based on the current YAML syntax. Can you check that? Otherwise this looks good.

**Comment by bdice** [comment]
@b-butler FYI, h5py has a new release 3.6.0 that supports Python 3.10 with cross-platform wheels. We can probably undo the special changes in this PR that were meant to work around that.

Release notes: https://docs.h5py.org/en/latest/whatsnew/3.6.html#new-features
Wheels: https://pypi.org/project/h5py/#files

**Comment by bdice** [comment]
This is now held by `tables` lacking Python 3.10 packages. Tracking issue here: https://github.com/PyTables/PyTables/issues/909

File: changelog.txt [hunk 1]
```diff
@@ -13,6 +13,7 @@ Version 1
 Added
 +++++
 
+ - Official support for Python 3.10. Not all optional dependencies are available for 3.10 yet (#631).
```
[the remark this answers is not in the document]
Comment (bdice) [reply changelog.txt]: I applied this suggestion but I am leaving this conversation open for visibility.

**Review by bdice — APPROVED** [review]
I approve this as-is, with the changelog notice that PyTables is not yet available for Python 3.10. I do not expect this to be a significant limitation for our users (pandas storage in H5Stores is uncommon to the best of my knowledge).  However, we may need to revisit our testing strategy to run with optional dependencies _except for PyTables_ if a PyTables release for Python 3.10 does not come out in the near future. This can be considered at a later time.

**Comment by bdice** [comment]
PyTables is available for Python 3.10 on conda-forge but its developers may stop creating wheel releases. https://github.com/PyTables/PyTables/issues/909#issuecomment-988992694

**Comment by bdice** [comment]
[PyTables 3.7.0](https://pypi.org/project/tables/3.7.0/#files) has been released with wheels for a wide range of platforms and Python versions. See: https://github.com/PyTables/PyTables/issues/909

I will finish up this PR and merge it.

**Pull request merged**